src/backend/vulkan/src/lib.rs

```diff
@@ -872,50 +886,43 @@ impl adapter::PhysicalDevice<Backend> for PhysicalDevice {
-        let memory_types = mem_properties.memory_types
-            [.. mem_properties.memory_type_count as usize]
+        let mut memory_types = Vec::new();
+        for mem in mem_properties
+            .memory_types[.. mem_properties.memory_type_count as usize]
             .iter()
-            .map(|mem| {
-                use crate::memory::Properties;
-                let mut type_flags = Properties::empty();
-
-                if mem
-                    .property_flags
-                    .intersects(vk::MemoryPropertyFlags::DEVICE_LOCAL)
-                {
-                    type_flags |= Properties::DEVICE_LOCAL;
-                }
-                if mem
-                    .property_flags
-                    .intersects(vk::MemoryPropertyFlags::HOST_VISIBLE)
-                {
-                    type_flags |= Properties::CPU_VISIBLE;
-                }
-                if mem
-                    .property_flags
-                    .intersects(vk::MemoryPropertyFlags::HOST_COHERENT)
-                {
-                    type_flags |= Properties::COHERENT;
-                }
-                if mem
-                    .property_flags
-                    .intersects(vk::MemoryPropertyFlags::HOST_CACHED)
-                {
-                    type_flags |= Properties::CPU_CACHED;
-                }
-                if mem
-                    .property_flags
-                    .intersects(vk::MemoryPropertyFlags::LAZILY_ALLOCATED)
-                {
-                    type_flags |= Properties::LAZILY_ALLOCATED;
-                }
+        {
+            use crate::memory::Properties;
+            let mut type_flags = Properties::empty();
+            let mut vk_flags = mem.property_flags;
 
-                adapter::MemoryType {
-                    properties: type_flags,
-                    heap_index: mem.heap_index as usize,
-                }
-            })
-            .collect();
+            if vk_flags.take(vk::MemoryPropertyFlags::DEVICE_LOCAL) {
+                type_flags |= Properties::DEVICE_LOCAL;
+            }
+            if vk_flags.take(vk::MemoryPropertyFlags::HOST_VISIBLE) {
+                type_flags |= Properties::CPU_VISIBLE;
+            }
+            if vk_flags.take(vk::MemoryPropertyFlags::HOST_COHERENT) {
+                type_flags |= Properties::COHERENT;
+            }
+            if vk_flags.take(vk::MemoryPropertyFlags::HOST_CACHED) {
+                type_flags |= Properties::CPU_CACHED;
+            }
+            if vk_flags.take(vk::MemoryPropertyFlags::LAZILY_ALLOCATED) {
+                type_flags |= Properties::LAZILY_ALLOCATED;
+            }
+
+            if !vk_flags.is_empty() {
+                warn!("Stopping memory type iteration on unknown flags {:?}", vk_flags);
+                // Vulkan guarantees that all the fancy memory types come after
+                // the non-fancy ones, so we can stop here.
+                break
+            }
```

Comment:
This condition is incorrect. [§10.2](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#memory-device):

> There is no ordering requirement between [any two elements of memoryTypes]  for the case their propertyFlags members are not in a subset relation.

An unknown memory property flag may appear at any position in the sequence, so long as it does so in a type whose properties are not a superset of those of a type that appears later. For example, the very first memory type could be a type with exactly one unknown property, causing this loop to unconditionally fail.

Instead, how about `.iter().filter(|x| SUPPORTED_FLAGS.contains(x.property_flags))`?

Reply:
I see. I was hoping we can get away with just taking the first N. Since once we start skipping elements, we'll need to introduce the bidirectional mapping inside a device, to know how user IDs map to Vulkan IDs and visa versa

Reply:
I don't have much context for what you're trying to do here, but `VK_MAX_MEMORY_TYPES` is only 32, so a bidirectional lookup table shouldn't be a big deal.

Reply:
Yeah, it's not really a big deal. Just something I was hoping to avoid :)